lib/evaluate/shape.cc

```diff
@@ -451,21 +468,80 @@ auto GetShapeHelper::operator()(const ProcedureRef &call) const -> Result {
+      // Reduction with DIM=
+      if (call.arguments().size() >= 2) {
+        auto arrayShape{
+            (*this)(UnwrapExpr<Expr<SomeType>>(call.arguments().at(0)))};
+        const auto *dimArg{UnwrapExpr<Expr<SomeType>>(call.arguments().at(1))};
+        if (arrayShape.has_value() && dimArg != nullptr) {
+          if (auto dim{ToInt64(*dimArg)}) {
+            if (*dim >= 1 &&
+                static_cast<std::size_t>(*dim) <= arrayShape->size()) {
+              arrayShape->erase(arrayShape->begin() + (*dim - 1));
+              return std::move(*arrayShape);
+            }
+          }
+        }
+      }
+    } else if (intrinsic->name == "cshift" || intrinsic->name == "eoshift") {
+      if (!call.arguments().empty()) {
+        return (*this)(call.arguments()[0]);
+      }
     } else if (intrinsic->name == "reshape") {
       if (call.arguments().size() >= 2 && call.arguments().at(1).has_value()) {
         // SHAPE(RESHAPE(array,shape)) -> shape
-        const auto *shapeExpr{call.arguments().at(1).value().UnwrapExpr()};
-        CHECK(shapeExpr != nullptr);
-        Expr<SomeInteger> shape{std::get<Expr<SomeInteger>>(shapeExpr->u)};
-        return AsShape(context_, ConvertToType<ExtentType>(std::move(shape)));
+        if (const auto *shapeExpr{
+                call.arguments().at(1).value().UnwrapExpr()}) {
+          auto shape{std::get<Expr<SomeInteger>>(shapeExpr->u)};
+          return AsShape(context_, ConvertToType<ExtentType>(std::move(shape)));
+        }
+      }
+    } else if (intrinsic->name == "spread") {
+      // SHAPE(SPREAD(ARRAY,DIM,NCOPIES)) = SHAPE(ARRAY) with NCOPIES inserted
+      // at position DIM.
+      if (call.arguments().size() == 3) {
+        auto arrayShape{
+            (*this)(UnwrapExpr<Expr<SomeType>>(call.arguments().at(0)))};
+        const auto *dimArg{UnwrapExpr<Expr<SomeType>>(call.arguments().at(1))};
+        const auto *nCopies{
+            UnwrapExpr<Expr<SomeInteger>>(call.arguments().at(2))};
+        if (arrayShape.has_value() && dimArg != nullptr && nCopies != nullptr) {
+          if (auto dim{ToInt64(*dimArg)}) {
+            if (*dim >= 1 &&
+                static_cast<std::size_t>(*dim) <= arrayShape->size() + 1) {
+              arrayShape->emplace(arrayShape->begin() + *dim - 1,
+                  ConvertToType<ExtentType>(common::Clone(*nCopies)));
+              return std::move(*arrayShape);
+            }
+          }
+        }
+      }
+    } else if (intrinsic->name == "transpose") {
+      if (call.arguments().size() >= 1) {
+        if (auto shape{(*this)(call.arguments().at(0))}) {
+          if (shape->size() == 2) {
+            std::swap((*shape)[0], (*shape)[1]);
+            return shape;
+          }
+        }
       }
+    } else if (intrinsic->characteristics.value().attrs.test(characteristics::
+                       Procedure::Attr::NullPointer)) {  // NULL(MOLD=)
+      return (*this)(call.arguments());
```

Comment:
Should there be curly braces around this "return"?

Reply:
They are present.